### ISSUE TITLE
Long-lived ZC commands can block JSON-RPC thread

### ISSUE THREAD

**Issue by bitcartel** [zcash/zcash#934]
Branch zc.v0.11.2.z2

A few seconds after invoking zcash-cli zcrawreceive, I decided to CTRL-C to check the block count via zcash-cli getblockcount.  This operation blocked.  So did getinfo.

Only after about 30-45 seconds (the amount of time zcrawreceive takes to run on my system) were commands being processed.

It would apepar that zcrawreceive was still running on the JSON-RPC thread even after I had closed the connection.

**Comment by ebfull** [comment]
This is a problem with our RPC in many places, but it's most critical when we need to cancel JoinSplit creation. We need to interrupt libsnark's prover somehow.

**Comment by nathan-at-least** [comment]
Since the new high-level API will have an async interface, interrupting an RPC request handling thread should become a much lower priority, on the theory that the long-running operations will delegate the work to a separate thread quickly in every case, and then reply and terminate.

Let's do this: after implementing the RPC interface with async operations, review existing remaining API calls to see if any are still worst-case-long-running and see what we can do at that point.

**Comment by nathan-at-least** [comment]
I'm closing this as "wontfix". If I'm missing something we can revive it.
